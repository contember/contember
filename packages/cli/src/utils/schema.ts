import { SchemaValidator, SchemaValidatorSkippedErrors, ValidationError } from '@contember/schema-utils'
import { Schema } from '@contember/schema'

export const printValidationErrors = (errors: ValidationError[], message?: string) => {
	if (errors.length === 0) {
		return
	}
	console.group(message || 'Schema is invalid:')
	for (const err of errors) {
		console.log(`${err.path.join('.')}: [${err.code}] ${err.message}`)
	}
	console.groupEnd()
}

export const validateSchemaAndPrintErrors = (schema: Schema, message?: string, skippedErrors?: SchemaValidatorSkippedErrors[]): boolean => {
	const errors = SchemaValidator.validate(schema, skippedErrors)
	printValidationErrors(errors, message)
	return errors.length === 0
}
