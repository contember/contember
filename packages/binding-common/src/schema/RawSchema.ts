import type { RawSchemaEntity } from './RawSchemaEntity'
import type { RawSchemaEnum } from './RawSchemaEnum'

export interface RawSchema {
	enums: RawSchemaEnum[]
	entities: RawSchemaEntity[]
}
