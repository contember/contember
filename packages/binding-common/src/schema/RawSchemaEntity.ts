import type { EntityName } from '../treeParameters'
import type { SchemaColumn } from './SchemaColumn'
import type { SchemaRelation } from './SchemaRelation'

export type RawSchemaFields = Array<SchemaRelation | SchemaColumn>

export interface RawSchemaEntity {
	name: EntityName
	customPrimaryAllowed: boolean
	fields: RawSchemaFields
	unique: { fields: string[] }[]
}
