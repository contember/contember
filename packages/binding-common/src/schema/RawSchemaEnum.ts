import type { SchemaEnumName } from './SchemaEnums'

export interface RawSchemaEnum {
	name: SchemaEnumName
	values: string[]
}
