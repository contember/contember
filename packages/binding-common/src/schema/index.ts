export * from './RawSchema'
export * from './Schema'
export * from './SchemaColumn'
export * from './SchemaColumnType'
export * from './SchemaEntities'
export * from './SchemaEntity'
export * from './SchemaEnums'
export * from './SchemaField'
export * from './SchemaFields'
export * from './SchemaLoader'
export * from './SchemaPreprocessor'
export * from './SchemaRelation'
export * from './SchemaRelationOrderBy'
export * from './SchemaStore'
export * from './SchemaUniqueConstraint'
