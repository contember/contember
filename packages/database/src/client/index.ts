export * from './Client'
export * from './errorCodes'
export * from './errors'
export * from './Connection'
export * from './EventManager'
export * from './Pool'
export * from './Transaction'
export * from './SingleConnection'
