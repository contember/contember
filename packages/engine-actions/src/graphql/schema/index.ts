import { GraphQLResolveInfo, GraphQLScalarType, GraphQLScalarTypeConfig } from 'graphql'
export type Maybe<T> = T | null
export type InputMaybe<T> = Maybe<T>
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] }
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> }
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> }
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never }
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never }
export type RequireFields<T, K extends keyof T> = Omit<T, K> & { [P in K]-?: NonNullable<T[P]> }
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
	ID: { input: string; output: string }
	String: { input: string; output: string }
	Boolean: { input: boolean; output: boolean }
	Int: { input: number; output: number }
	Float: { input: number; output: number }
	DateTime: { input: Date; output: Date }
	Json: { input: any; output: any }
	Uuid: { input: string; output: string }
}

export type Event = {
	readonly __typename?: 'Event'
	readonly createdAt: Scalars['DateTime']['output']
	readonly id: Scalars['Uuid']['output']
	readonly identityId?: Maybe<Scalars['Uuid']['output']>
	readonly ipAddress?: Maybe<Scalars['String']['output']>
	readonly lastStateChange: Scalars['DateTime']['output']
	readonly log: Scalars['Json']['output']
	readonly numRetries: Scalars['Int']['output']
	readonly payload: Scalars['Json']['output']
	readonly resolvedAt?: Maybe<Scalars['DateTime']['output']>
	readonly stage: Scalars['String']['output']
	readonly state: EventState
	readonly target: Scalars['String']['output']
	readonly transactionId: Scalars['Uuid']['output']
	readonly userAgent?: Maybe<Scalars['String']['output']>
	readonly visibleAt?: Maybe<Scalars['DateTime']['output']>
}

export type EventArgs = {
	/** Max 10000 */
	readonly limit?: InputMaybe<Scalars['Int']['input']>
	readonly offset?: InputMaybe<Scalars['Int']['input']>
}

export type EventState =
  | 'created'
  | 'failed'
  | 'processing'
  | 'retrying'
  | 'stopped'
  | 'succeed'

export type Mutation = {
	readonly __typename?: 'Mutation'
	readonly processBatch: ProcessBatchResponse
	readonly retryEvent: RetryEventResponse
	readonly setVariables: SetVariablesResponse
	readonly stopEvent: StopEventResponse
}


export type MutationRetryEventArgs = {
	id: Scalars['Uuid']['input']
}


export type MutationSetVariablesArgs = {
	args: SetVariablesArgs
}


export type MutationStopEventArgs = {
	id: Scalars['Uuid']['input']
}

export type ProcessBatchResponse = {
	readonly __typename?: 'ProcessBatchResponse'
	readonly ok: Scalars['Boolean']['output']
}

export type Query = {
	readonly __typename?: 'Query'
	readonly event?: Maybe<Event>
	readonly eventsInProcessing: ReadonlyArray<Event>
	readonly eventsToProcess: ReadonlyArray<Event>
	readonly failedEvents: ReadonlyArray<Event>
	readonly variables: ReadonlyArray<Variable>
}


export type QueryEventArgs = {
	id: Scalars['Uuid']['input']
}


export type QueryEventsInProcessingArgs = {
	args?: InputMaybe<EventArgs>
}


export type QueryEventsToProcessArgs = {
	args?: InputMaybe<EventArgs>
}


export type QueryFailedEventsArgs = {
	args?: InputMaybe<EventArgs>
}

export type RetryEventResponse = {
	readonly __typename?: 'RetryEventResponse'
	readonly ok: Scalars['Boolean']['output']
}

export type SetVariablesArgs = {
	readonly mode?: InputMaybe<SetVariablesMode>
	readonly variables: ReadonlyArray<VariableInput>
}

/**
 * Defines how it handles original variables.
 * - MERGE merges with new values (default behaviour)
 * - SET replaces all variables
 * - APPEND_ONLY_MISSING appends values if not already exist
 */
export type SetVariablesMode =
  | 'APPEND_ONLY_MISSING'
  | 'MERGE'
  | 'SET'

export type SetVariablesResponse = {
	readonly __typename?: 'SetVariablesResponse'
	readonly ok: Scalars['Boolean']['output']
}

export type StopEventResponse = {
	readonly __typename?: 'StopEventResponse'
	readonly ok: Scalars['Boolean']['output']
}

export type Variable = {
	readonly __typename?: 'Variable'
	readonly name: Scalars['String']['output']
	readonly value: Scalars['String']['output']
}

export type VariableInput = {
	readonly name: Scalars['String']['input']
	readonly value: Scalars['String']['input']
}



export type ResolverTypeWrapper<T> = Promise<T> | T


export type ResolverWithResolve<TResult, TParent, TContext, TArgs> = {
	resolve: ResolverFn<TResult, TParent, TContext, TArgs>
}
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> = ResolverFn<TResult, TParent, TContext, TArgs> | ResolverWithResolve<TResult, TParent, TContext, TArgs>

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
	parent: TParent,
	args: TArgs,
	context: TContext,
	info: GraphQLResolveInfo
) => Promise<TResult> | TResult

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
	parent: TParent,
	args: TArgs,
	context: TContext,
	info: GraphQLResolveInfo
) => AsyncIterable<TResult> | Promise<AsyncIterable<TResult>>

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
	parent: TParent,
	args: TArgs,
	context: TContext,
	info: GraphQLResolveInfo
) => TResult | Promise<TResult>

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
	subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>
	resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
	subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>
	resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
	parent: TParent,
	context: TContext,
	info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (obj: T, context: TContext, info: GraphQLResolveInfo) => boolean | Promise<boolean>

export type NextResolverFn<T> = () => Promise<T>

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
	next: NextResolverFn<TResult>,
	parent: TParent,
	args: TArgs,
	context: TContext,
	info: GraphQLResolveInfo
) => TResult | Promise<TResult>



/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = {
	Boolean: ResolverTypeWrapper<Scalars['Boolean']['output']>
	DateTime: ResolverTypeWrapper<Scalars['DateTime']['output']>
	Event: ResolverTypeWrapper<Event>
	EventArgs: EventArgs
	EventState: EventState
	Int: ResolverTypeWrapper<Scalars['Int']['output']>
	Json: ResolverTypeWrapper<Scalars['Json']['output']>
	Mutation: ResolverTypeWrapper<{}>
	ProcessBatchResponse: ResolverTypeWrapper<ProcessBatchResponse>
	Query: ResolverTypeWrapper<{}>
	RetryEventResponse: ResolverTypeWrapper<RetryEventResponse>
	SetVariablesArgs: SetVariablesArgs
	SetVariablesMode: SetVariablesMode
	SetVariablesResponse: ResolverTypeWrapper<SetVariablesResponse>
	StopEventResponse: ResolverTypeWrapper<StopEventResponse>
	String: ResolverTypeWrapper<Scalars['String']['output']>
	Uuid: ResolverTypeWrapper<Scalars['Uuid']['output']>
	Variable: ResolverTypeWrapper<Variable>
	VariableInput: VariableInput
}

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = {
	Boolean: Scalars['Boolean']['output']
	DateTime: Scalars['DateTime']['output']
	Event: Event
	EventArgs: EventArgs
	Int: Scalars['Int']['output']
	Json: Scalars['Json']['output']
	Mutation: {}
	ProcessBatchResponse: ProcessBatchResponse
	Query: {}
	RetryEventResponse: RetryEventResponse
	SetVariablesArgs: SetVariablesArgs
	SetVariablesResponse: SetVariablesResponse
	StopEventResponse: StopEventResponse
	String: Scalars['String']['output']
	Uuid: Scalars['Uuid']['output']
	Variable: Variable
	VariableInput: VariableInput
}

export interface DateTimeScalarConfig extends GraphQLScalarTypeConfig<ResolversTypes['DateTime'], any> {
	name: 'DateTime'
}

export type EventResolvers<ContextType = any, ParentType extends ResolversParentTypes['Event'] = ResolversParentTypes['Event']> = {
	createdAt?: Resolver<ResolversTypes['DateTime'], ParentType, ContextType>
	id?: Resolver<ResolversTypes['Uuid'], ParentType, ContextType>
	identityId?: Resolver<Maybe<ResolversTypes['Uuid']>, ParentType, ContextType>
	ipAddress?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>
	lastStateChange?: Resolver<ResolversTypes['DateTime'], ParentType, ContextType>
	log?: Resolver<ResolversTypes['Json'], ParentType, ContextType>
	numRetries?: Resolver<ResolversTypes['Int'], ParentType, ContextType>
	payload?: Resolver<ResolversTypes['Json'], ParentType, ContextType>
	resolvedAt?: Resolver<Maybe<ResolversTypes['DateTime']>, ParentType, ContextType>
	stage?: Resolver<ResolversTypes['String'], ParentType, ContextType>
	state?: Resolver<ResolversTypes['EventState'], ParentType, ContextType>
	target?: Resolver<ResolversTypes['String'], ParentType, ContextType>
	transactionId?: Resolver<ResolversTypes['Uuid'], ParentType, ContextType>
	userAgent?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>
	visibleAt?: Resolver<Maybe<ResolversTypes['DateTime']>, ParentType, ContextType>
	__isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>
}

export interface JsonScalarConfig extends GraphQLScalarTypeConfig<ResolversTypes['Json'], any> {
	name: 'Json'
}

export type MutationResolvers<ContextType = any, ParentType extends ResolversParentTypes['Mutation'] = ResolversParentTypes['Mutation']> = {
	processBatch?: Resolver<ResolversTypes['ProcessBatchResponse'], ParentType, ContextType>
	retryEvent?: Resolver<ResolversTypes['RetryEventResponse'], ParentType, ContextType, RequireFields<MutationRetryEventArgs, 'id'>>
	setVariables?: Resolver<ResolversTypes['SetVariablesResponse'], ParentType, ContextType, RequireFields<MutationSetVariablesArgs, 'args'>>
	stopEvent?: Resolver<ResolversTypes['StopEventResponse'], ParentType, ContextType, RequireFields<MutationStopEventArgs, 'id'>>
}

export type ProcessBatchResponseResolvers<ContextType = any, ParentType extends ResolversParentTypes['ProcessBatchResponse'] = ResolversParentTypes['ProcessBatchResponse']> = {
	ok?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>
	__isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>
}

export type QueryResolvers<ContextType = any, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = {
	event?: Resolver<Maybe<ResolversTypes['Event']>, ParentType, ContextType, RequireFields<QueryEventArgs, 'id'>>
	eventsInProcessing?: Resolver<ReadonlyArray<ResolversTypes['Event']>, ParentType, ContextType, Partial<QueryEventsInProcessingArgs>>
	eventsToProcess?: Resolver<ReadonlyArray<ResolversTypes['Event']>, ParentType, ContextType, Partial<QueryEventsToProcessArgs>>
	failedEvents?: Resolver<ReadonlyArray<ResolversTypes['Event']>, ParentType, ContextType, Partial<QueryFailedEventsArgs>>
	variables?: Resolver<ReadonlyArray<ResolversTypes['Variable']>, ParentType, ContextType>
}

export type RetryEventResponseResolvers<ContextType = any, ParentType extends ResolversParentTypes['RetryEventResponse'] = ResolversParentTypes['RetryEventResponse']> = {
	ok?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>
	__isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>
}

export type SetVariablesResponseResolvers<ContextType = any, ParentType extends ResolversParentTypes['SetVariablesResponse'] = ResolversParentTypes['SetVariablesResponse']> = {
	ok?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>
	__isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>
}

export type StopEventResponseResolvers<ContextType = any, ParentType extends ResolversParentTypes['StopEventResponse'] = ResolversParentTypes['StopEventResponse']> = {
	ok?: Resolver<ResolversTypes['Boolean'], ParentType, ContextType>
	__isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>
}

export interface UuidScalarConfig extends GraphQLScalarTypeConfig<ResolversTypes['Uuid'], any> {
	name: 'Uuid'
}

export type VariableResolvers<ContextType = any, ParentType extends ResolversParentTypes['Variable'] = ResolversParentTypes['Variable']> = {
	name?: Resolver<ResolversTypes['String'], ParentType, ContextType>
	value?: Resolver<ResolversTypes['String'], ParentType, ContextType>
	__isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>
}

export type Resolvers<ContextType = any> = {
	DateTime?: GraphQLScalarType
	Event?: EventResolvers<ContextType>
	Json?: GraphQLScalarType
	Mutation?: MutationResolvers<ContextType>
	ProcessBatchResponse?: ProcessBatchResponseResolvers<ContextType>
	Query?: QueryResolvers<ContextType>
	RetryEventResponse?: RetryEventResponseResolvers<ContextType>
	SetVariablesResponse?: SetVariablesResponseResolvers<ContextType>
	StopEventResponse?: StopEventResponseResolvers<ContextType>
	Uuid?: GraphQLScalarType
	Variable?: VariableResolvers<ContextType>
}

