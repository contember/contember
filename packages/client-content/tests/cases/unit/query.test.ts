import { describe, expect, test } from 'bun:test'
import { createClient } from '../../lib'
import { Input } from '@contember/schema'
import OrderDirection = Input.OrderDirection
import { queryBuilder } from '../../client'

describe('queries', () => {

	const qb = queryBuilder
	test('list', async () => {
		const [client, calls] = createClient({
			authors: [
				{
					id: '1',
					name: 'John',
					email: 'foo@localhost',
				},
				{
					id: '2',
					name: 'John',
					email: 'bar@localhost',
				},
			],
		})
		const result = await client.query({
			authors: qb.list('Author', {}, it => it.$$()),
		})
		expect(result).toStrictEqual({
			authors: [
				{
					id: '1',
					name: 'John',
					email: 'foo@localhost',
				},
				{
					id: '2',
					name: 'John',
					email: 'bar@localhost',
				},
			],
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toStrictEqual({})
	})

	test('multiple queries', async () => {
		const [client, calls] = createClient({
			authors: [
				{
					name: 'John',
					email: 'foo@localhost',
				},
				{
					name: 'John',
					email: 'bar@localhost',
				},
			],
			posts: [
				{
					title: 'Post 1',
					content: 'Content 1',
				},
				{
					title: 'Post 2',
					content: 'Content 2',
				},
			],
		})
		const result = await client.query({
			authors: qb.list('Author', {}, it => it.$$()),
			posts: qb.list('Post', {}, it => it.$$()),
		})
		expect(result as any).toStrictEqual({
			authors: [
				{
					name: 'John',
					email: 'foo@localhost',
				},
				{
					name: 'John',
					email: 'bar@localhost',
				},
			],
			posts: [
				{
					title: 'Post 1',
					content: 'Content 1',
				},
				{
					title: 'Post 2',
					content: 'Content 2',
				},
			],
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toEqual({})
	})


	test('single query', async () => {
		const [client, calls] = createClient({
			value: [
				{
					title: 'Post 1',
					content: 'Content 1',
				},
				{
					title: 'Post 2',
					content: 'Content 2',
				},
			],
		})

		const result = await client.query(qb.list('Post', {}, it => it.$$()))
		expect(result as any).toStrictEqual([
			{
				title: 'Post 1',
				content: 'Content 1',
			},
			{
				title: 'Post 2',
				content: 'Content 2',
			},
		])
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toEqual({})
	})


	test('nested object', async () => {
		const [client, calls] = createClient()
		await client.query({
			authors: qb.list('Author', {}, it => it.$$().$('posts', {}, it => it.$$().$('tags', {}, it => it.$$()))),
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toEqual({})
	})
	test('nested object args', async () => {
		const [client, calls] = createClient()
		await client.query({
			authors: qb.list('Author', {}, it => it.$$().$('posts', {
				limit: 10,
				filter: { tags: { name: { eq: 'foo' } } },
			}, it => it.$$())),
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toStrictEqual({
			'Int_1': 10,
			'PostWhere_0': {
				'tags': {
					'name': {
						'eq': 'foo',
					},
				},
			},
		})
	})


	test('list with args', async () => {
		const [client, calls] = createClient()
		await client.query({
			authors: qb.list('Author', {
				filter: { name: { eq: 'John' } },
				orderBy: [{ name: OrderDirection.asc }],
				limit: 10,
				offset: 20,
			}, it => it.$$()),
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot(`
			"query($AuthorWhere_0: AuthorWhere, $AuthorOrderBy_1: [AuthorOrderBy!], $Int_2: Int, $Int_3: Int) {
				authors: listAuthor(filter: $AuthorWhere_0, orderBy: $AuthorOrderBy_1, limit: $Int_2, offset: $Int_3) {
					id
					name
					email
				}
			}
			"
		`)
		expect(calls[0].variables).toStrictEqual(
			{
				'AuthorOrderBy_1': [
					{
						'name': 'asc',
					},
				],
				'AuthorWhere_0': {
					'name': {
						'eq': 'John',
					},
				},
				'Int_2': 10,
				'Int_3': 20,
			},
		)
	})


	test('get by id', async () => {
		const [client, calls] = createClient()
		const result = await client.query({
			authors: qb.get('Author', {
				by: { id: 'ca7a9b84-efbb-435d-a063-da11f205335a' },
			}, it => it.$$()),
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toEqual({
			'AuthorUniqueWhere_0': {
				'id': 'ca7a9b84-efbb-435d-a063-da11f205335a',
			},
		})
	})

	test('top level list transform', async () => {
		const [client, calls] = createClient({
			authors: [
				{
					name: 'John',
					email: 'foo@localhost',
				},
				{
					name: 'Jane',
					email: 'bar@localhost',
				},
			],
		})
		const result = await client.query({
			authors: qb.list('Author', {}, it => it
				.$$()
				.transform(it => ({
					...it,
					label: `${it.name} (${it.email})`,
				})),
			),
		})
		expect(result as any).toStrictEqual({
			authors: [
				{
					name: 'John',
					email: 'foo@localhost',
					label: 'John (foo@localhost)',
				},
				{
					name: 'Jane',
					email: 'bar@localhost',
					label: 'Jane (bar@localhost)',
				},
			],
		})
		expect(calls).toHaveLength(1)
		expect(calls).toMatchSnapshot()
	})

	test('top level get transform', async () => {
		const [client, calls] = createClient({
			author: {
				name: 'John',
				email: 'foo@localhost',
			},
		})
		const result = await client.query({
			author: qb.get('Author', { by: { id: '123' } }, it => it
				.$$()
				.transform(it => ({
					...it,
					label: `${it.name} (${it.email})`,
				})),
			),
		})
		expect(result as any).toStrictEqual({
			author: {
				name: 'John',
				email: 'foo@localhost',
				label: 'John (foo@localhost)',
			},
		})
		expect(calls).toHaveLength(1)
		expect(calls).toMatchSnapshot()
	})


	test('nested has-many transform', async () => {
		const [client, calls] = createClient({
			author: {
				name: 'John',
				email: 'foo@localhost',
				posts: [
					{
						publishedAt: '2021-01-01T00:00:00Z',
					},
					{
						publishedAt: '2021-01-02T00:00:00Z',
					},
				],
			},
		})
		const result = await client.query({
			author: qb.get('Author', { by: { id: '123' } }, it => it
				.$$()
				.$('posts', {}, it => it.$$().transform(it => ({
					...it,
					publishYear: it.publishedAt ? new Date(it.publishedAt).getFullYear() : null,
				}))),
			),
		})
		expect(result as any).toStrictEqual({
			author: {
				name: 'John',
				email: 'foo@localhost',
				posts: [
					{
						publishedAt: '2021-01-01T00:00:00Z',
						publishYear: 2021,
					},
					{
						publishedAt: '2021-01-02T00:00:00Z',
						publishYear: 2021,
					},
				],
			},
		})
		expect(calls).toHaveLength(1)
		expect(calls).toMatchSnapshot()
	})


	test('nested has-one transform', async () => {
		const [client, calls] = createClient({
			post: {
				publishedAt: '2021-01-01T00:00:00Z',
				author: {
					name: 'John',
					email: 'foo@localhost',
				},
			},
		})
		const author = qb.fragment('Author', it => it
			.$$()
			.transform(it => ({
				...it,
				label: `${it.name} (${it.email})`,
			})),
		)

		const result = await client.query({
			post: qb.get('Post', { by: { id: '123' } }, it => it
				.$$()
				.$('author', author),
			),
		})
		expect(result as any).toStrictEqual({
			post: {
				publishedAt: '2021-01-01T00:00:00Z',
				author: {
					name: 'John',
					email: 'foo@localhost',
					label: 'John (foo@localhost)',
				},
			},
		})
		expect(calls).toMatchSnapshot()
	})


	test('nested has-many-by transform', async () => {
		const [client, calls] = createClient({
			post: {
				publishedAt: '2021-01-01T00:00:00Z',
				localesByLocale: {
					title: 'Hello',
				},
			},
		})
		const result = await client.query({
			post: qb.get('Post', { by: { id: '123' } }, it => it
				.$$()
				.$('localesByLocale', { by: { locale: { code: 'en' } } }, it => it.$$().transform(it => ({
					...it,
					label: it.title?.toUpperCase(),
				}))),
			),
		})
		expect(result as any).toStrictEqual({
			post: {
				publishedAt: '2021-01-01T00:00:00Z',
				localesByLocale: {
					title: 'Hello',
					label: 'HELLO',
				},
			},
		})
		expect(calls).toMatchSnapshot()
	})

	test('multiple transforms', async () => {
		const [client, calls] = createClient({
			author: {
				name: 'John',
				email: 'foo@localhost',
				posts: [
					{
						publishedAt: '2021-01-01T00:00:00Z',
					},
					{
						publishedAt: '2021-01-02T00:00:00Z',
					},
				],
			},
		})
		const result = await client.query({
			author: qb.get('Author', { by: { id: '123' } }, it => it
				.$$()
				.transform(it => ({
					...it,
					value1: 'foo',
				}))
				.$('posts', {}, it => it.$$().transform(it => ({
					...it,
					postLabel: 'foo bar',
				})))
				.transform(it => ({
					...it,
					value2: 'bar',
				})),
			),
		})
		expect(result as any).toStrictEqual({
			author: {
				name: 'John',
				email: 'foo@localhost',
				value1: 'foo',
				value2: 'bar',
				posts: [
					{
						publishedAt: '2021-01-01T00:00:00Z',
						postLabel: 'foo bar',
					},
					{
						publishedAt: '2021-01-02T00:00:00Z',
						postLabel: 'foo bar',
					},
				],
			},
		})
		expect(calls).toHaveLength(1)
	})

	test('omit', async () => {
		const [client, calls] = createClient({
			authors: [
				{
					id: '1',
					name: 'John',
				},
			],
		})
		const result = await client.query({
			authors: qb.list('Author', {}, it => it.$$().omit('email').$('posts', {}, it => it.$$()).omit('posts')),
		})
		expect(result).toStrictEqual({
			authors: [
				{
					id: '1',
					name: 'John',
				},
			],
		})
		expect(calls).toHaveLength(1)
		expect(calls[0].query).toMatchSnapshot()
		expect(calls[0].variables).toStrictEqual({})
	})

})
