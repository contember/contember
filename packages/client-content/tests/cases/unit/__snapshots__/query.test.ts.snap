// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`queries list 1`] = `
"query {
\tauthors: listAuthor {
\t\tid
\t\tname
\t\temail
\t}
}
"
`;

exports[`queries multiple queries 1`] = `
"query {
\tauthors: listAuthor {
\t\tid
\t\tname
\t\temail
\t}
\tposts: listPost {
\t\tid
\t\tpublishedAt
\t}
}
"
`;

exports[`queries single query 1`] = `
"query {
\tvalue: listPost {
\t\tid
\t\tpublishedAt
\t}
}
"
`;

exports[`queries nested object 1`] = `
"query {
\tauthors: listAuthor {
\t\tid
\t\tname
\t\temail
\t\tposts {
\t\t\tid
\t\t\tpublishedAt
\t\t\ttags {
\t\t\t\tid
\t\t\t\tname
\t\t\t}
\t\t}
\t}
}
"
`;

exports[`queries nested object args 1`] = `
"query($PostWhere_0: PostWhere, $Int_1: Int) {
\tauthors: listAuthor {
\t\tid
\t\tname
\t\temail
\t\tposts(filter: $PostWhere_0, limit: $Int_1) {
\t\t\tid
\t\t\tpublishedAt
\t\t}
\t}
}
"
`;

exports[`queries list with args: 
\t\t\t"query($AuthorWhere_0: AuthorWhere, $AuthorOrderBy_1: [AuthorOrderBy!], $Int_2: Int, $Int_3: Int) {
\t\t\t\tauthors: listAuthor(filter: $AuthorWhere_0, orderBy: $AuthorOrderBy_1, limit: $Int_2, offset: $Int_3) {
\t\t\t\t\tid
\t\t\t\t\tname
\t\t\t\t\temail
\t\t\t\t}
\t\t\t}
\t\t\t"
\t\t 1`] = `
"query($AuthorWhere_0: AuthorWhere, $AuthorOrderBy_1: [AuthorOrderBy!], $Int_2: Int, $Int_3: Int) {
\tauthors: listAuthor(filter: $AuthorWhere_0, orderBy: $AuthorOrderBy_1, limit: $Int_2, offset: $Int_3) {
\t\tid
\t\tname
\t\temail
\t}
}
"
`;

exports[`queries get by id 1`] = `
"query($AuthorUniqueWhere_0: AuthorUniqueWhere!) {
\tauthors: getAuthor(by: $AuthorUniqueWhere_0) {
\t\tid
\t\tname
\t\temail
\t}
}
"
`;

exports[`queries nested has-many transform 1`] = `
[
  {
    "query": 
"query($AuthorUniqueWhere_0: AuthorUniqueWhere!) {
	author: getAuthor(by: $AuthorUniqueWhere_0) {
		id
		name
		email
		posts {
			id
			publishedAt
		}
	}
}
"
,
    "variables": {
      "AuthorUniqueWhere_0": {
        "id": "123",
      },
    },
  },
]
`;

exports[`queries top level list transform 1`] = `
[
  {
    "query": 
"query {
	authors: listAuthor {
		id
		name
		email
	}
}
"
,
    "variables": {},
  },
]
`;

exports[`queries top level get transform 1`] = `
[
  {
    "query": 
"query($AuthorUniqueWhere_0: AuthorUniqueWhere!) {
	author: getAuthor(by: $AuthorUniqueWhere_0) {
		id
		name
		email
	}
}
"
,
    "variables": {
      "AuthorUniqueWhere_0": {
        "id": "123",
      },
    },
  },
]
`;

exports[`queries nested has-one transform 1`] = `
[
  {
    "query": 
"query($PostUniqueWhere_0: PostUniqueWhere!) {
	post: getPost(by: $PostUniqueWhere_0) {
		id
		publishedAt
		author {
			id
			name
			email
		}
	}
}
"
,
    "variables": {
      "PostUniqueWhere_0": {
        "id": "123",
      },
    },
  },
]
`;

exports[`queries nested has-many-by transform 1`] = `
[
  {
    "query": 
"query($PostUniqueWhere_0: PostUniqueWhere!, $PostLocalesByLocaleUniqueWhere_1: PostLocalesByLocaleUniqueWhere!) {
	post: getPost(by: $PostUniqueWhere_0) {
		id
		publishedAt
		localesByLocale(by: $PostLocalesByLocaleUniqueWhere_1) {
			id
			title
			content
		}
	}
}
"
,
    "variables": {
      "PostLocalesByLocaleUniqueWhere_1": {
        "locale": {
          "code": "en",
        },
      },
      "PostUniqueWhere_0": {
        "id": "123",
      },
    },
  },
]
`;

exports[`queries omit 1`] = `
"query {
	authors: listAuthor {
		id
		name
	}
}
"
`;
